standalone/src/main/scala/io/delta/standalone/internal/ConflictChecker.scala

```diff
@@ -0,0 +1,161 @@
+  val protocol: Seq[Protocol] = actions.collect { case a: Protocol => a }
+  val commitInfo: Option[CommitInfo] = actions.collectFirst { case a: CommitInfo => a }.map(
+    ci => ci.copy(version = Some(commitVersion)))
+  val removedFiles: Seq[RemoveFile] = actions.collect { case a: RemoveFile => a }
+  val addedFiles: Seq[AddFile] = actions.collect { case a: AddFile => a }
+  val isBlindAppendOption: Option[Boolean] = commitInfo.flatMap(_.isBlindAppend)
+  val blindAppendAddedFiles: Seq[AddFile] = if (isBlindAppendOption.getOrElse(false)) {
+    addedFiles
+  } else {
+    Seq()
+  }
+  val changedDataAddedFiles: Seq[AddFile] = if (isBlindAppendOption.getOrElse(false)) {
+    Seq()
+  } else {
+    addedFiles
+  }
+  val onlyAddFiles: Boolean = actions.collect { case f: FileAction => f }
+    .forall(_.isInstanceOf[AddFile])
+}
+
+private[internal] class ConflictChecker(
+    currentTransactionInfo: CurrentTransactionInfo,
+    winningCommitVersion: Long,
+    isolationLevel: IsolationLevel) {
+
+  private val winningCommitSummary: WinningCommitSummary = createWinningCommitSummary()
+
+  def checkConflicts(): Unit = {
+//    TODO checkProtocolCompatibility()
+//    TODO checkNoMetadataUpdates()
+    checkForAddedFilesThatShouldHaveBeenReadByCurrentTxn()
+    checkForDeletedFilesAgainstCurrentTxnReadFiles()
+//    TODO checkForDeletedFilesAgainstCurrentTxnDeletedFiles()
+//    TODO checkForUpdatedApplicationTransactionIdsThatCurrentTxnDependsOn()
+  }
+
+  /**
+   * Initializes [[WinningCommitSummary]] for the already committed
+   * transaction (winning transaction).
+   */
+  private def createWinningCommitSummary(): WinningCommitSummary = {
+      val deltaLog = currentTransactionInfo.deltaLog
+      val winningCommitActions = deltaLog.store.read(
+        FileNames.deltaFile(deltaLog.logPath, winningCommitVersion)).map(Action.fromJson)
+      WinningCommitSummary(winningCommitActions, winningCommitVersion)
+  }
+
+  /**
+   * Check if the new files added by the already committed transactions should have been read by
+   * the current transaction.
+   */
+  def checkForAddedFilesThatShouldHaveBeenReadByCurrentTxn(): Unit = {
+    // Fail if new files have been added that the txn should have read.
+    val addedFilesToCheckForConflicts = isolationLevel match {
+      case Serializable =>
+        winningCommitSummary.changedDataAddedFiles ++ winningCommitSummary.blindAppendAddedFiles
+      case SnapshotIsolation =>
+        Seq.empty
+    }
+
+    val doesConflict = currentTransactionInfo.externalConflictChecker
+      .doesConflict(addedFilesToCheckForConflicts.map(ConversionUtils.convertAddFile).asJava)
+
+    if (doesConflict) {
+      throw DeltaErrors.concurrentAppendException(winningCommitSummary.commitInfo)
+    }
+  }
+
+  /**
+   * Check if [[RemoveFile]] actions added by already committed transactions conflicts with files
+   * read by the current transaction.
+   */
+  def checkForDeletedFilesAgainstCurrentTxnReadFiles(): Unit = {
```

Comment:
TODO what about the default scenario? We assume that the current txn read the entire table. So readFiles will be empty, but this should still fail.